libs/langchain/langchain/schema/runnable/locals.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+from typing import Any, AsyncIterator, Dict, Iterator, Mapping, Optional, Union
+
+from langchain.load.serializable import Serializable
+from langchain.schema.runnable import Runnable, RunnableConfig, RunnablePassthrough
+from langchain.schema.runnable.base import Input, Output
+
+
+class PutLocalVar(RunnablePassthrough):
+    key: Union[str, Mapping[str, str]]
+    """The key(s) to use for storing the input variable(s) in local state.
+    
+    If a string is provided then the entire input is stored under that key. If a 
+        Mapping is provided, then the map values are gotten from the input and 
+        stored in local state under the map keys.
+    """
+
+    def __init__(self, key: str, **kwargs: Any) -> None:
+        super().__init__(key=key, **kwargs)
+
+    def _put(self, input: Input, *, config: Optional[RunnableConfig] = None) -> None:
+        if config is None:
+            raise ValueError(
+                "PutLocalVar should only be used in a RunnableSequence, and should "
+                "therefore always receive a non-null config."
+            )
+        if isinstance(self.key, str):
+            config["_locals"][self.key] = input
+        elif isinstance(input, Mapping):
+            for input_key, put_key in self.key.items():
+                config["_locals"][put_key] = input[input_key]
+        else:
+            raise TypeError(
+                f"`key` should be a string or Mapping[str, str], received type "
+                f"{(type(self.key))}."
+            )
+
+    def _concat_put(
+        self, input: Input, *, config: Optional[RunnableConfig] = None
+    ) -> None:
+        if config is None:
+            raise ValueError(
+                "PutLocalVar should only be used in a RunnableSequence, and should "
+                "therefore always receive a non-null config."
+            )
+        if isinstance(self.key, str):
+            if self.key not in config["_locals"]:
+                config["_locals"][self.key] = input
+            else:
+                config["_locals"][self.key] += input
+        elif isinstance(input, Mapping):
+            for input_key, put_key in self.key.items():
+                if put_key not in config["_locals"]:
+                    config["_locals"][put_key] = input
+                else:
+                    config["_locals"][put_key] += input
+        else:
+            raise TypeError(
+                f"`key` should be a string or Mapping[str, str], received type "
+                f"{(type(self.key))}."
+            )
+
+    def invoke(self, input: Input, config: Optional[RunnableConfig] = None) -> Input:
+        self._put(input, config=config)
+        return super().invoke(input, config)
+
+    async def ainvoke(
+        self, input: Input, config: RunnableConfig | None = None
+    ) -> Input:
+        self._put(input, config=config)
+        return await super().ainvoke(input, config)
+
+    def transform(
```

Comment:
do we want to worry about transform / is concatenating state the behavior we want or should we collect the whole input @nfcampos